File: concurrency-model-6.1.x/retry/package-repository/server-config-java/src/java/src/com/example/serverconfig/serverconfigRFS.java
package com.example.serverconfig;

import com.example.serverconfig.namespaces.*;
import com.tailf.conf.*;
import com.tailf.dp.*;
import com.tailf.dp.annotations.*;
import com.tailf.dp.proto.*;
import com.tailf.dp.services.*;
import com.tailf.maapi.Maapi;
import com.tailf.maapi.MaapiException;
import com.tailf.maapi.MaapiRetryableOp;
import com.tailf.maapi.MaapiUserSessionFlag;
import com.tailf.navu.*;
import com.tailf.ncs.NcsMain;
import com.tailf.ncs.ns.Ncs;
import com.tailf.ncs.template.Template;
import com.tailf.ncs.template.TemplateVariables;
import java.io.IOException;
import java.lang.StringBuilder;
import java.net.InetAddress;
import java.net.Socket;
import java.util.List;
import java.util.Properties;
import java.util.concurrent.Semaphore;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class serverconfigRFS {
  private static Logger log = LogManager.getLogger(serverconfigRFS.class);
  private String deviceName;
  private String serverName;
  private Semaphore sem = new Semaphore(1);

  /**
   * Create callback method.
   * This method is called when a service instance committed due to a create
   * or update event.
   *
   * This method returns a opaque as a Properties object that can be null.
   * If not null it is stored persistently by Ncs.
   * This object is then delivered as argument to new calls of the create
   * method for this service (fastmap algorithm).
   * This way the user can store and later modify persistent data outside
   * the service model that might be needed.
   *
   * @param context - The current ServiceContext object
   * @param service - The NavuNode references the service node.
   * @param ncsRoot - This NavuNode references the ncs root.
   * @param opaque  - Parameter contains a Properties object.
   *                  This object may be used to transfer
   *                  additional information between consecutive
   *                  calls to the create callback.  It is always
   *                  null in the first call. I.e. when the service
   *                  is first created.
   * @return Properties the returning opaque instance
   * @throws ConfException
   */

  @ServiceCallback(servicePoint="dns-config-servicepoint",
                   callType=ServiceCBType.CREATE)
  public Properties createdns(ServiceContext context,
                              NavuNode service,
                              NavuNode ncsRoot,
                              Properties opaque)
      throws ConfException {

    Template dnsTemplate = new Template(context, "dns-config-template");
    TemplateVariables dnsVars = new TemplateVariables();
    try {
      log.info("Service create(service="
               + service.getKeyPath().toString() + ")");
      String serverName = service.leaf("server").value().toString();
      String serverIp = ncsRoot.getParent()//
                               .container(serverConfig.uri)//
                               .container("servers").list("server")//
                               .elem(serverName).leaf("ip").value()//
                               .toString();
      /* Acquire a seemaphore which only purpose is to enable simulating
         a config update conflict before continuing */
      log.info("DNS service - try acquiring the semaphore");
      sem.acquire();
      dnsVars.putQuoted("DNS_IP", serverIp);
      // apply the template with the variable
      dnsTemplate.apply(service, dnsVars);
      log.info("Service create done(service="
               + service.getKeyPath().toString() + ")");
      sem.release();
    } catch (Exception e) {
      throw new DpCallbackException(e.getMessage(), e);
    }
    return opaque;
  }

  @ServiceCallback(servicePoint="ntp-config-servicepoint",
                   callType=ServiceCBType.CREATE)
  public Properties createntp(ServiceContext context,
                              NavuNode service,
                              NavuNode ncsRoot,
                              Properties opaque)
      throws ConfException {
    try {
      log.info("Service create(service="
               + service.getKeyPath().toString() + ")");
      String serverName = service.leaf("server").value().toString();
      String serverIp = ncsRoot.getParent().container(serverConfig.uri)//
                               .container("servers").list("server")//
                               .elem(serverName).leaf("ip")//
                               .value().toString();

      /* The below transaction will cause a conflict and trigger an
         automatic retry */
      Socket socket = new Socket(NcsMain.getInstance().getNcsHost(),
                                 NcsMain.getInstance().getNcsPort());
      Maapi maapi = new Maapi(socket);
      maapi.startUserSession("admin",
                             InetAddress.getByName(null),
                             "example",
                             new String[] {"admin"},
                             MaapiUserSessionFlag.PROTO_TCP);
      NavuContext ncontext = new NavuContext(maapi);
      ncontext.startRunningTrans(Conf.MODE_READ_WRITE);
      NavuContainer root = new NavuContainer(ncontext);
      NavuContainer scRoot = root.container(new serverConfig().hash());
      NavuLeaf ipLeaf = scRoot.container("servers").list("server")//
                              .elem(serverName).leaf("ip");
      int start = serverIp.lastIndexOf(".");
      StringBuilder builder = new StringBuilder();
      builder.append(serverIp.substring(0, start));
      builder.append(".121");
      serverIp = builder.toString();
      ipLeaf.set(serverIp);
      ncontext.applyClearTrans();
      socket.close();

      TemplateVariables ntpVars = new TemplateVariables();
      ntpVars.putQuoted("NTP_IP", serverIp);
      // apply the template with the variable
      Template ntpTemplate = new Template(context,
                                          "ntp-config-template");
      ntpTemplate.apply(service, ntpVars);
      log.info("Service create done(service="
               + service.getKeyPath().toString() + ")");
    } catch (Exception e) {
      throw new DpCallbackException(e.getMessage(), e);
    }
    return opaque;
  }

  /**
   * Acquire semaphore action
   */
  @ActionCallback(callPoint="acquire-sem", callType=ActionCBType.ACTION)
  public ConfXMLParam[] acquiresem(DpActionTrans trans, ConfTag name,
                                   ConfObject[] kp, ConfXMLParam[] params)
      throws DpCallbackException {
    try {
      log.info("Acquire action");
      sem.acquire();
      return new ConfXMLParam[] {
        new ConfXMLParamValue("sc", "result", new ConfBool(true))};
    } catch (Exception e) {
      throw new DpCallbackException("Acquire semaphore action failed",
                                    e);
    }
  }

  /**
   * Release semaphore action
   */
  @ActionCallback(callPoint="release-sem", callType=ActionCBType.ACTION)
  public ConfXMLParam[] releasesem(DpActionTrans trans, ConfTag name,
                                   ConfObject[] kp, ConfXMLParam[] params)
      throws DpCallbackException {
    try {
      log.info("Release action");
      sem.release();
      return new ConfXMLParam[] {
        new ConfXMLParamValue("sc", "result", new ConfBool(true))};
    } catch (Exception e) {
      throw new DpCallbackException("Release semaphore action failed",
                                    e);
    }
  }

  /**
   * Update NTP action
   */
  @ActionCallback(callPoint="update-ntp", callType=ActionCBType.ACTION)
  public ConfXMLParam[] updatentp(DpActionTrans trans, ConfTag name,
                                  ConfObject[] kp, ConfXMLParam[] params)
      throws DpCallbackException {
    try {
      String deviceName = params[0].getValue().toString();
      String serverName = params[1].getValue().toString();
      Socket socket = new Socket(NcsMain.getInstance().getNcsHost(),
                                 NcsMain.getInstance().getNcsPort());
      Maapi maapi = new Maapi(socket);
      maapi.startUserSession("admin",
                             InetAddress.getByName(null),
                             "example",
                             new String[] {"admin"},
                             MaapiUserSessionFlag.PROTO_TCP);
      NavuContext context = new NavuContext(maapi);
      context.startRunningTrans(Conf.MODE_READ_WRITE);
      NavuContainer root = new NavuContainer(context);
      NavuContainer scRoot = root.container(new serverConfig().hash());
      String serverIp = scRoot.container("servers").list("server")//
                              .elem(serverName).leaf("ip").value().toString();

      log.info("Update NTP without retry action - try acquiring the"
               + " semaphore");
      sem.acquire();
      NavuContainer ncsRoot = root.container(new Ncs().hash());
      ncsRoot.container("devices").list("device").elem(deviceName)//
             .container("config").container("sys").container("ntp")//
             .list("server").create(serverIp);
      try {
        context.applyClearTrans();
      } catch (NavuException e) {
        log.error(e);
      }
      sem.release();
      log.info("Update NTP without retry action - done");
      socket.close();
    } catch (Exception e) {
      throw new DpCallbackException("Update NTP action failed", e);
    }
    return new ConfXMLParam[] {
      new ConfXMLParamValue("sc", "result", new ConfBool(true))};
  }

  /**
   * Update NTP with retry action
   */
  @ActionCallback(callPoint="update-ntp-retry", callType=ActionCBType.ACTION)
  public ConfXMLParam[] updatentpretry(DpActionTrans trans, ConfTag name,
                                       ConfObject[] kp, ConfXMLParam[] params)
      throws DpCallbackException {
    try {
      deviceName = params[0].getValue().toString();
      serverName = params[1].getValue().toString();
      Socket socket = new Socket(NcsMain.getInstance().getNcsHost(),
                                 NcsMain.getInstance().getNcsPort());
      Maapi maapi = new Maapi(socket);
      maapi.startUserSession("admin",
                             InetAddress.getByName(null),
                             "example",
                             new String[]{"admin"},
                             MaapiUserSessionFlag.PROTO_TCP);
      log.info("Update NTP with retry action");
      maapi.ncsRunWithRetry(new MyProvisioningOp());
      log.info("Update NTP with retry action - finish");
      maapi.endUserSession();
      socket.close();
      return new ConfXMLParam[] {
        new ConfXMLParamValue("sc", "result", new ConfBool(true))};
    } catch (Exception e) {
      throw new DpCallbackException("Update NTP action failed", e);
    }
  }

  class MyProvisioningOp implements MaapiRetryableOp {
    public boolean execute(Maapi maapi, int tid) throws IOException,
                           ConfException, MaapiException {
      NavuContext context = new NavuContext(maapi, tid);
      NavuContainer root = new NavuContainer(context);
      NavuContainer scRoot = root.container(new serverConfig().hash());
      String serverIp = scRoot.container("servers").list("server")//
                              .elem(serverName).leaf("ip").value().toString();

      log.info("Update NTP with retry action - try acquiring the"
               + " semaphore");
      try {
        sem.acquire();
        NavuContainer ncsRoot = root.container(new Ncs().hash());
        ncsRoot.container("devices").list("device").elem(deviceName)//
               .container("config").container("sys").container("ntp")//
               .list("server").create(serverIp);
        sem.release();
      } catch (InterruptedException e) {
        throw new ConfException("Retry NTP server config update failed",
                                e);
      }
      log.info("Update NTP with retry action - done");
      return true;
    }
  }
}

File: concurrency-model-6.0.x/retry/package-repository/server-config-py/python/server_config/serverconfigRFS.py
# -*- mode: python; python-indent: 4 -*-
"""NSO concurrency model conflict retry example.

Implement services and actions with retry on transaction conflicts
(C) 2022 Tail-f Systems
Permission to use this code as a starting point hereby granted

See the README file for more information
"""
from multiprocessing import Manager
import ncs
from ncs.application import Service
from ncs.dp import Action


# -------------------------
# SERVICE CALLBACK EXAMPLES
# -------------------------
class DNSServiceCallbacks(Service):
    '''DNS server service'''
    def init(self, init_args):
        self.server_semaphore = init_args[0]

    @Service.create
    def cb_create(self, tctx, root, service, proplist):
        '''DNS server service create callback'''
        self.log.info('Service create(service=', service._path, ')')
        server_ip = root.servers.server[service.server].ip
        # Acquire a seemaphore which only purpose is to enable simulating a
        # config update conflict before continuing
        self.log.info('DNS service - try acquiring the semaphore')
        with self.server_semaphore:
            template_vars = ncs.template.Variables()
            template_vars.add("DNS_IP", server_ip)
            template = ncs.template.Template(service)
            template.apply("dns-config-template", template_vars)
        self.log.info('Service create done(service=', service._path, ')')


class NTPServiceCallbacks(Service):
    '''NTP server service'''
    @Service.create
    def cb_create(self, tctx, root, service, proplist):
        '''NTP server service create callback'''
        self.log.info('Service create(service=', service._path, ')')
        server_ip = root.servers.server[service.server].ip

        # The below transaction will cause a conflict and trigger an automatic
        # retry
        server_ip = server_ip[:server_ip.rfind('.')+1] + "121"
        with ncs.maapi.single_write_trans('admin', 'python') as trans:
            mroot = ncs.maagic.get_root(trans)
            mroot.servers.server[service.server].ip = server_ip
            trans.apply()

        template_vars = ncs.template.Variables()
        template_vars.add("NTP_IP", server_ip)
        template = ncs.template.Template(service)
        template.apply("ntp-config-template", template_vars)
        self.log.info('Service create done(service=', service._path, ')')


# ---------------
# ACTION EXAMPLES
# ---------------
class AcquireActionHandler(Action):
    '''Acquire action handler'''
    def init(self, init_args):
        self.server_semaphore = init_args[0]

    @Action.action
    def cb_action(self, uinfo, name, keypath, ainput, aoutput):
        '''Action callback'''
        self.log.info("Acquire action")
        self.server_semaphore.acquire()
        aoutput.result = True


class ReleaseActionHandler(Action):
    '''Release action handler'''
    def init(self, init_args):
        self.server_semaphore = init_args[0]

    @Action.action
    def cb_action(self, uinfo, name, keypath, ainput, aoutput):
        '''Action callback'''
        self.log.info("Release action")
        self.server_semaphore.release()
        aoutput.result = True


class UpdateNTPActionHandler(Action):
    '''Update NTP action handler'''
    def init(self, init_args):
        self.server_semaphore = init_args[0]

    @Action.action
    def cb_action(self, uinfo, name, keypath, ainput, aoutput):
        '''Action callback'''
        # Acquire a semaphore whose only purpose is to enable simulating a
        # config update conflict before continuing
        with ncs.maapi.single_write_trans('admin', 'python') as trans:
            root = ncs.maagic.get_root(trans)
            server_ip = root.servers.server[ainput.server].ip
            self.log.info("Update NTP without retry action - try acquiring" +
                          " the semaphore")
            with self.server_semaphore:
                ntp = root.devices.device[ainput.device].config.sys.ntp
                ntp.server.create(server_ip)
                trans.apply()
        self.log.info("Update NTP without retry action - done")
        aoutput.result = True


class UpdateNTPRetryActionHandler(Action):
    '''Update NTP with retry action handler'''
    def init(self, init_args):
        self.server_semaphore = init_args[0]

    @Action.action
    def cb_action(self, uinfo, name, keypath, ainput, aoutput):
        '''Action callback'''
        def update_ntp(trans, device, server):
            root = ncs.maagic.get_root(trans)
            server_ip = root.servers.server[ainput.server].ip
            # Acquire a semaphore which only purpose is to enable simulating a
            # config update conflict before continuing
            self.log.info("Update NTP with retry action - try acquiring" +
                          " the semaphore")
            with self.server_semaphore:
                ntp = root.devices.device[ainput.device].config.sys.ntp
                ntp.server.create(server_ip)
            self.log.info("Update NTP with retry action - done")
            return True

        self.log.info("Update NTP with retry action")
        with ncs.maapi.Maapi() as maapi:
            with ncs.maapi.Session(maapi, 'admin', 'python'):
                maapi.run_with_retry(lambda t: update_ntp(t, ainput.device,
                                     ainput.server))
        aoutput.result = True


class UpdateNTPRetryDecActionHandler(Action):
    '''Update NTP with retry decorator action handler'''
    def init(self, init_args):
        self.server_semaphore = init_args[0]

    @Action.action
    def cb_action(self, uinfo, name, keypath, ainput, aoutput):
        '''Action callback'''
        @ncs.maapi.retry_on_conflict()
        def update_ntp(device, server):
            with ncs.maapi.single_write_trans('admin', 'python') as trans:
                root = ncs.maagic.get_root(trans)
                server_ip = root.servers.server[ainput.server].ip
                # Acquire a semaphore which only purpose is to enable
                # simulating a config update conflict before continuing
                self.log.info("Update NTP with decorator retry action - try" +
                              " acquiring the semaphore")
                with self.server_semaphore:
                    ntp = root.devices.device[ainput.device].config.sys.ntp
                    ntp.server.create(server_ip)
                trans.apply()
                self.log.info("Update NTP with retry decorator action - done")
            return True

        self.log.info("Update NTP with retry decorator action")
        aoutput.result = update_ntp(ainput.device, ainput.server)


# --------------------------------------------
# COMPONENT THREAD THAT WILL BE STARTED BY NSO
# --------------------------------------------
class ServerApp(ncs.application.Application):
    '''Concurrency model retry service and action callbacks on conflict'''
    def setup(self):
        # The application class sets up logging for us. It is accessible
        # through 'self.log' and is a ncs.log.Log instance.
        self.log.info('Main RUNNING')

        manager = Manager()
        server_semaphore = manager.Semaphore()
        init_args = [server_semaphore]

        # Service callbacks require a registration for a 'service point',
        # as specified in the corresponding data model.
        self.register_service('dns-config-servicepoint', DNSServiceCallbacks,
                              init_args)
        self.register_service('ntp-config-servicepoint', NTPServiceCallbacks)

        # Action callback registration
        self.register_action('acquire-sem', AcquireActionHandler, init_args)
        self.register_action('release-sem', ReleaseActionHandler, init_args)
        self.register_action('update-ntp', UpdateNTPActionHandler, init_args)
        self.register_action('update-ntp-retry', UpdateNTPRetryActionHandler,
                             init_args)
        self.register_action('update-ntp-retry-dec',
                             UpdateNTPRetryDecActionHandler, init_args)

        # If we registered any callback(s) above, the Application class
        # took care of creating a daemon (related to the service/action point).

        # When this setup method is finished, all registrations are
        # considered done and the application is 'started'.

    def teardown(self):
        # When the application is finished (which would happen if NCS went
        # down, packages were reloaded or some error occurred) this teardown
        # method will be called.

        self.log.info('Main FINISHED')
